README
A C++ wrapper for fam from sgi.
fam can be found at http://oss.sgi.com/projects/fam/

fam uses imon to inform it when inodes change, the net 
effect being that applications can register interest
in a file, and have events send to the application 
when the file changes.

The fam++ wrapper (hopefully) makes fam a little easier
to use. It does this by using exceptions instead of 
return values to indicate error, allow type safe
callbacks using libsig++ and compile time checking.

This is version 2 of fam++ which is non compatible with
version 1. This is due to moving to loki and a cleanup
of the API. The general use of fam++ will change in
this version to hopefully make it simpler to use.


ChangeLog
6.0.1

2009-10-16  monkeyiq  <monkeyiq@users.sourceforge.net>

	* updates for sigc++ 2.2
	
5.1.1

2006-03-22  monkeyiq  <monkeyiq@users.sourceforge.net>

	* Released 5.1.0
	  Optional ability to build against libsigc++-2.x
	
5.1.0

2006-03-22  monkeyiq  <monkeyiq@users.sourceforge.net>

	* Optional ability to build against libsigc++-2.x

2005-11-09  monkeyiq  <monkeyiq@users.sourceforge.net>

	* released 5.0.0
	
5.0.0

2005-11-09  monkeyiq  <monkeyiq@users.sourceforge.net>

	* Moved to using ferrisloki for more consistent smartpointer handling.
	
4.0.0

2005-09-06  monkeyiq  <monkeyiq@users.sourceforge.net>

	* Made STLport optional
	
(3.5.2)
	
2005-02-03  monkeyiq  <monkeyiq@users.sourceforge.net>

	* cleanups to pkg-config file

2004-05-31  monkeyiq  <monkeyiq@users.sourceforge.net>

	* Updates for gcc 3.4 compilation

(3.5.0)
	Mon Aug 18, 2003 @  6:34 PM
	Changed some of the factory use to avoid ICE in gcc 3.3+

(3.3.1)
	Mon May 19, 2003 @ 12:33 PM
	Moved to using libsigc++ 1.2 and some dist changes for rh9

Sat Feb 23 01:57:36 EST 2002 (2.2.1)
	* added more includes to headers to make them more self contained. fixed exception specs to make gcc3
          happier, though I have not done a 100% gcc3 compile/test yet.


Fri Sep 28 03:27:52 EST 2001
	* Changed the lib to only need glib2, not gtk2


Sat Sep  8 03:06:19 EST 2001 (2.1.0)
	* Improved some exception specs with throw()
	* Ported to STLport 4.0


Sun Aug 26 20:44:40 EST 2001 (VERSION2 2.0.0)
	
	* Moved to using loki.
	* Cleaned up code

-----------------------------------------------------------


27-Oct-2000 Ben Martin  (VERSION 0.0.3)

	* Added FamSingleton class to header. See gpdial_applet for an example of usage (check freshmeat).
	* Added getInstance_gnome_and_wait_user_wants_to_quit exception and getInstance_gnome_and_wait().
		getInstance_gnome_and_wait will keep polling for fam to be available, and also show
		a dialog box so that the user can close the application if fam does not become available.

23-Oct-2000 Ben Martin

	* Created famstatic.hh after I started using this wrapper in
	  anger. I found that there were issues with non template functions
	  being multiply defined. Old code will need to include famstatic.hh
	  now. I don't think anyone is using fam++ at the moment though :-0

	* Updated the examples to use famstatic.hh too.


9-July-2000 Ben Martin

	* Fixed bug where fampp would keep a stale reference to
		each FamppRequest even after the FamppRequest was deleted().

	* Fixed bug where after FamppRequest lookup in fampp no check
		was made if the request was == null. (possible segv was there).

	* Added GPL COPYING and header comments to all cpp and header
		files indicating GPL.

	* Created a README

	* Created this file


src/Fampp2.cpp
/******************************************************************************
*******************************************************************************
*******************************************************************************

    Copyright (C) 2001 Ben Martin

    This file is part of fampp2.

    fampp2 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    fampp2 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with fampp2.  If not, see <http://www.gnu.org/licenses/>.

    For more details see the COPYING file in the root directory of this
    distribution.

    $Id: Fampp2.cpp,v 1.5 2010/09/24 05:16:50 ben Exp $

*******************************************************************************
*******************************************************************************
******************************************************************************/

#include <Fampp2.hh>
#include <Fampp2_private.hh>
#include <iostream>

#include <sigc++/sigc++.h>
#include <sigc++/slot.h>
#include <sigc++/object.h>
#include <sigc++/object_slot.h>


using namespace std;


namespace Fampp
{
    FamppSingletonClass& FamppInstance()
    {
        return Fampp::Instance();
    }
    FAMEvent* getCurrentFAMEvent()
    {
        FamppInstance().getFAMEvent();
    }
    

    static int
    FamReqNum()
    {
        const FAMEvent* ev = Fampp::Instance().getFAMEvent();
        return FAMREQUEST_GETREQNUM( (&(ev->fr)) );
    }

    int
    FamReqNum( FamppRequest *req )
    {
        return FAMREQUEST_GETREQNUM( &req->getRequest() );
    }

    static int
    FamReqNum( fh_fampp_req req )
    {
        return FamReqNum( GetImpl( req ) );
    }




    FamppRequest::FamppRequest(
        const string& filename,
        void* userData,
        bool isFile )
    {
        if(isFile)
        {
            if( 0 != FAMMonitorFile( 
                    &getFAMConnection(),
                    filename.c_str(),
                    &theFAMRequest,
                    userData)) 
            {
                throw FamppFileMonitorInitFailedException(filename);
            }
        }
        else
        {
            if( 0 != FAMMonitorDirectory( 
                    &getFAMConnection(),
                    filename.c_str(),
                    &theFAMRequest,
                    userData)) 
            {
                throw FamppDirMonitorInitFailedException(filename);
            }
        }
    }

// #include <unistd.h>

    FamppRequest::~FamppRequest()
    {
//         cerr << "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!" << endl;
//         cerr << "FamppRequest::~FamppRequest() this:" << (void*)this << endl;
//         cerr << "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!" << endl;
//         sleep(5);
        
        if( 0 != FAMCancelMonitor(
                &getFAMConnection(), 
                &theFAMRequest)) 
        {
            throw FamppRequestCancelFailedException(this);
        }
        Fampp::Instance().removeRequest(this);
    }

    FAMRequest&
    FamppRequest::getRequest()
    {
        return theFAMRequest;
    }

    fh_fampp_req
    FamppRequest::toHandle()
    {
//         cerr << "FamppRequest::toHandle(enter)" << endl;
        fh_fampp_req req = fh_fampp_req(this);
//        req->AddRef();
//         cerr << "FamppRequest::toHandle(exit)" << endl;
        return req;
    }
    

    void
    FamppRequest::dispatch()
    {
//        cerr << "FamppRequest::dispatch(enter)" << endl;
        
        FamppEventBase* ev = EventFactory::Instance().CreateObject( getFamEvent()->code );
        
        {
//             cerr << "FamppRequest::dispatch(1)" << endl;
            const fh_fampp_req& req = toHandle();
//             cerr << "FamppRequest::dispatch(2)" << endl;
            ev->dispatch( req );
//             cerr << "FamppRequest::dispatch(3)" << endl;
        }
        
        
//         Sig.emit( getFAMEvent()->code,
//                   getFAMEvent()->getFileName(),
//                   this,
//                   getFAMEvent());

//        cerr << "FamppRequest::dispatch(exit)" << endl;
        
    }


    void
    FamppRequest::suspend()
    {
        if( 0 != FAMSuspendMonitor(
                &getFAMConnection(), 
                &getRequest())) 
        {
            throw FamppSuspendFailedException(this);
        }
    }

    void
    FamppRequest::resume()
    {
        if( 0 != FAMResumeMonitor(
                &getFAMConnection(), 
                &getRequest())) 
        {
            throw FamppResumeFailedException(this);
        }
    }

    FAMEvent*
    FamppRequest::getFamEvent()
    {
        return Fampp::Instance().getFAMEvent();
    }
    
    FAMConnection&
    FamppRequest::getFAMConnection()
    {
        return Fampp::Instance().getFAMConnection();
    }
    
    void
    FamppRequest::bounceEv(
        string filename,
        fh_fampp_req req,
        fh_fampp_ev ev )
    {
        getMuxedSig().emit( filename, req, ev );
    }
    
    FamppRequest::MuxedSig_t&
    FamppRequest::getMuxedSig()
    {
        static bool virgin = true;

        if( virgin )
        {
            virgin = false;
            
            getSig<FamppChangedEvent>().connect(sigc::mem_fun( *this, &FamppRequest::bounceEv));
//             getSig<FamppDeletedEvent>().connect(sigc::mem_fun( *this, &FamppRequest::bounceEv));
//             getSig<FamppStartExecutingEvent>().connect(sigc::mem_fun( this, &FamppRequest::bounceEv));
//             getSig<FamppStopExecutingEvent>().connect(sigc::mem_fun( this, &FamppRequest::bounceEv));
//             getSig<FamppCreatedEvent>().connect(sigc::mem_fun( this, &FamppRequest::bounceEv));
//             getSig<FamppMovedEvent>().connect(sigc::mem_fun( this, &FamppRequest::bounceEv));
//             getSig<FamppAcknowledgeEvent>().connect(sigc::mem_fun( this, &FamppRequest::bounceEv));
//             getSig<FamppExistsEvent>().connect(sigc::mem_fun( this, &FamppRequest::bounceEv));
//             getSig<FamppEndExistEvent>().connect(sigc::mem_fun( this, &FamppRequest::bounceEv));
        }
        return MuxedSig;
    }
    

    
    ///////////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////////


    FamppImplMethods::FamppImplMethods()
    {
        if( 0 != FAMOpen( &getFAMConnection()))
        {
            throw FamppOpenFailedException();
        }
    }

    
    FamppImplMethods::~FamppImplMethods()
    {
        if(0 != FAMClose( &theFAMConnection )) {
            throw FamppCloseFailedException();
        }
    }

    
    FAMConnection&
    FamppImplMethods::getFAMConnection()
    {
        return theFAMConnection;
    }
    
    int
    FamppImplMethods::getFD()
    {
        return FAMCONNECTION_GETFD((&theFAMConnection));
    }
    

    FAMEvent*
    FamppImplMethods::getFAMEvent()
    {
        return &theFAMEvent;
    }


    
    ///////////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////////


    FamppSingletonClass::FamppSingletonClass()
    {
    }

    FamppSingletonClass::FamppSingletonClass(const string appName)
    {
    }
    
    FamppSingletonClass::~FamppSingletonClass()
    {
    }

    fh_fampp_req
    FamppSingletonClass::makeFamppRequest(
        string filename, 
        void* userData,
        bool isFile) 
    {
        fh_fampp_req ret = new FamppRequest(filename,userData,isFile);
//        cerr << "FamppSingletonClass::makeFamppRequest()... rc:" << ret->getReferenceCount() << endl;
        return ret;
    }


    void
    FamppSingletonClass::removeRequest(FamppRequest *req)
    {
        FAMppRequestsByID[ FamReqNum( req ) ] = 0;
    }


    
    fh_fampp_req
    FamppSingletonClass::MonitorDirectory( string filename, void* userData ) 
        throw(
            FamppDirMonitorInitFailedException
            )
    {
        fh_fampp_req ret = makeFamppRequest( filename, userData, false );
//        cerr << "MonitorDirectory. filename:" << filename << " FamReqNum(ret):" << FamReqNum(ret) << endl;
        FAMppRequestsByID[ FamReqNum(ret) ] = GetImpl(ret);
        return ret;
    }
    

    fh_fampp_req
    FamppSingletonClass::MonitorFile( string filename, void* userData ) 
        throw (FamppFileMonitorInitFailedException)
    {
        fh_fampp_req ret = makeFamppRequest( filename, userData, true );
        FAMppRequestsByID[ FamReqNum(ret) ] = GetImpl(ret);
        return ret;
    }



    void
    FamppSingletonClass::NextEvent()
    {
        try
        {
            if( 0 > FAMNextEvent(
                    &getFAMConnection(), 
                    getFAMEvent())) 
            {
                throw FamppNextEventFailedException();
            }
            
            if( fh_fampp_req req = FAMppRequestsByID[ FamReqNum() ] )
            {
                req->dispatch();
            }
        }
        catch( exception& e )
        {
            addException( e );
        }
    }

    int
    FamppSingletonClass::Pending()
    {
        return FAMPending(&getFAMConnection());
    }

	
    
    ///////////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////////

    
    fh_fampp_req MonitorDirectory( string filename, void* userData ) 
        throw(FamppDirMonitorInitFailedException)
    {
        return Fampp::Instance().MonitorDirectory( filename, userData );
    }
    
        

    fh_fampp_req MonitorFile( string filename, void* userData ) 
        throw(FamppFileMonitorInitFailedException)
    {
        return Fampp::Instance().MonitorFile( filename, userData );
    }
    
    
    
};



src/FamppEvents.cpp
/******************************************************************************
*******************************************************************************
*******************************************************************************

    Copyright (C) 2001 Ben Martin

    This file is part of fampp2.

    fampp2 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    fampp2 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with fampp2.  If not, see <http://www.gnu.org/licenses/>.

    For more details see the COPYING file in the root directory of this
    distribution.

    $Id: FamppEvents.cpp,v 1.3 2010/09/24 05:16:51 ben Exp $

*******************************************************************************
*******************************************************************************
******************************************************************************/

#include <FamppEvents.hh>

using namespace std;


///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////

namespace Fampp
{
    

const string FamppChangedEventStr = "FAMChanged Event";
static FamppChangedEvent FamppChangedEventObj;

const string FamppDeletedEventStr = "FAMDeleted Event";
static FamppDeletedEvent FamppDeletedEventObj;

const string FamppStartExecutingEventStr = "FAMStartExecuting Event";
static FamppStartExecutingEvent FamppStartExecutingEventObj;

const string FamppStopExecutingEventStr = "FAMStopExecuting Event";
static FamppStopExecutingEvent FamppStopExecutingEventObj;

const string FamppCreatedEventStr = "FAMCreated Event";
static FamppCreatedEvent FamppCreatedEventObj;

const string FamppMovedEventStr = "FAMMoved Event";
static FamppMovedEvent FamppMovedEventObj;

const string FamppAcknowledgeEventStr = "FAMMoved Event";
static FamppAcknowledgeEvent FamppAcknowledgeEventObj;

const string FamppExistsEventStr = "FAMExists Event";
static FamppExistsEvent FamppExistsEventObj;

const string FamppEndExistEventStr = "FAMEndExist Event";
static FamppEndExistEvent FamppEndExistEventObj;

    struct RegisterEventClasses
    {

        template< class T >
        static FamppEventBase* CreateEvent()
            {
                FamppEventBase* ret = 0;
                ret = new T;
                return ret;
            }
        RegisterEventClasses()
            {
                Loki::Factory< FamppEventBase, int >& fac = EventFactory::Instance();

                fac.Register( FAMChanged,        &CreateEvent<FamppChangedEvent> );
                fac.Register( FAMDeleted,        &CreateEvent<FamppDeletedEvent> );
                fac.Register( FAMStartExecuting, &CreateEvent<FamppStartExecutingEvent> );
                fac.Register( FAMStopExecuting,  &CreateEvent<FamppStopExecutingEvent> );
                fac.Register( FAMCreated,        &CreateEvent<FamppCreatedEvent> );
                fac.Register( FAMMoved,          &CreateEvent<FamppMovedEvent> );
                fac.Register( FAMAcknowledge,    &CreateEvent<FamppAcknowledgeEvent> );
                fac.Register( FAMExists,         &CreateEvent<FamppExistsEvent> );
                fac.Register( FAMEndExist,       &CreateEvent<FamppEndExistEvent> );
            }
    };
    static RegisterEventClasses __RegisterEventClassesObject;
};


examples/gtk/famppgtk.cpp
/******************************************************************************
*******************************************************************************
*******************************************************************************

    Copyright (C) 2001 Ben Martin

    This file is part of fampp2.
    This is a simple gtk+ application that monitors /tmp using fam++

    fampp2 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    fampp2 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with fampp2.  If not, see <http://www.gnu.org/licenses/>.

    For more details see the COPYING file in the root directory of this
    distribution.

    $Id: famppgtk.cpp,v 1.4 2010/09/24 05:16:50 ben Exp $

*******************************************************************************
*******************************************************************************
******************************************************************************/

#include <sigc++/sigc++.h>
#include <sigc++/slot.h>
#include <sigc++/object.h>
#include <sigc++/object_slot.h>

#include <gtk/gtk.h>
#include <stdio.h>

#include <Fampp2.hh>
#include <Fampp2GlibSupport.hh>

#include <iostream>

using namespace Fampp;
using namespace std;



static gint delete_event_cb(GtkWidget* w, GdkEventAny* e, gpointer data);
static void button_click_cb(GtkWidget* w, gpointer data);


class ThisApp : public sigc::trackable
{
public:

    void CreatedFamEv(
        string filename,
        fh_fampp_req req,
        fh_fampp_ev ev )
        {
            cerr << "CreatedFamEv() " 
                 << " event:" << ev->getEventName() 
                 << " filename:" << filename
                 << endl;
        }


    void MuxedFamEv(
        string filename,
        fh_fampp_req req,
        fh_fampp_ev ev )
        {
            cerr << "MuxedFamEv() " 
                 << " event:" << ev->getEventName() 
                 << " filename:" << filename
                 << endl;
        }
    

};



int 
main(int argc, char* argv[])
{
    GtkWidget* window;
    GtkWidget* button;
    GtkWidget* label;
    ThisApp thisobj;
  

    gtk_init(&argc, &argv);  

    window = gtk_window_new(GTK_WINDOW_TOPLEVEL);

    button = gtk_button_new();

    label  = gtk_label_new("Change /tmp!");

    gtk_container_add(GTK_CONTAINER(button), label);
    gtk_container_add(GTK_CONTAINER(window), button);

    gtk_window_set_title(GTK_WINDOW(window), "Hello");
    gtk_container_set_border_width(GTK_CONTAINER(button), 10);
  
    gtk_signal_connect(GTK_OBJECT(window),
                       "delete_event",
                       GTK_SIGNAL_FUNC(delete_event_cb),
                       NULL);

    gtk_signal_connect(GTK_OBJECT(button),
                       "clicked",
                       GTK_SIGNAL_FUNC(button_click_cb),
                       label);

	
	try {

        ////////////////////////////////////////////////////////////
        ////////////////////////////////////////////////////////////
        ////////////////////////////////////////////////////////////
		// Begin FAM++ code                                       //
        ////////////////////////////////////////////////////////////
        ////////////////////////////////////////////////////////////
        ////////////////////////////////////////////////////////////

        // Hookup to glib for automatic handling during main loop
        RegisterFamppWithGLib();
        
        // Create a request for /tmp
        cerr << "===1======================================" << endl;
        fh_fampp_req req = MonitorDirectory("/tmp");
        cerr << "===2======================================" << endl;

        // Connect libsigC++ signal
        req->getSig<FamppCreatedEvent>();

        req->getSig<FamppCreatedEvent>().connect(sigc::mem_fun(thisobj, &ThisApp::CreatedFamEv));

        // Connect all signals to a method.
//        req->getMuxedSig().connect(SigC::slot(&thisobj, &ThisApp::MuxedFamEv));
        
        cerr << "===3======================================" << endl;

        
        
        ////////////////////////////////////////////////////////////
        ////////////////////////////////////////////////////////////
		// end of fam specific code, (fam) callbacks initiated by gtk_main() for us 
		// gtk only code from here!
        ////////////////////////////////////////////////////////////
        ////////////////////////////////////////////////////////////
        ////////////////////////////////////////////////////////////

        cerr << "=========================================" << endl;
        cerr << "=========================================" << endl;
        cerr << "=========================================" << endl;
        
		gtk_widget_show_all(window);
		gtk_main();
		return 0;
    }
	catch(FamppException e) {
		cout << e << endl;
	}
}

static gint 
delete_event_cb(GtkWidget* window, GdkEventAny* e, gpointer data)
{
    gtk_main_quit();
    return FALSE;
}

static void 
button_click_cb(GtkWidget* w, gpointer data)
{
    GtkWidget* label;
    gchar* text;
    gchar* tmp;

    label = GTK_WIDGET(data);
    gtk_label_get(GTK_LABEL(label), &text);
    tmp = g_strdup(text);
    g_strreverse(tmp);
    gtk_label_set_text(GTK_LABEL(label), tmp);
    g_free(tmp);
}


examples/console_test/famppct.cpp
/******************************************************************************
*******************************************************************************
*******************************************************************************

    Copyright (C) 2001 Ben Martin

    This file is part of fampp2.

    fampp2 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    fampp2 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with fampp2.  If not, see <http://www.gnu.org/licenses/>.

    For more details see the COPYING file in the root directory of this
    distribution.

    $Id: famppct.cpp,v 1.5 2010/09/24 05:16:50 ben Exp $

*******************************************************************************
*******************************************************************************
******************************************************************************/

#include <sigc++/sigc++.h>
#include <sigc++/slot.h>
#include <sigc++/object.h>
#include <sigc++/object_slot.h>

#include <glib.h>
#include <stdio.h>

#include <Fampp2.hh>
#include <Fampp2GlibSupport.hh>

#include <iostream>

using namespace std;


void CreatedFamEv( string filename, Fampp::fh_fampp_req req, Fampp::fh_fampp_ev ev )
{
    cerr << "CreatedFamEv() " 
         << " event:" << ev->getEventName() 
         << " filename:" << filename
         << endl;
}


void
OnFamppExistsEvent( string filename, Fampp::fh_fampp_req req, Fampp::fh_fampp_ev ev )
{
    cerr << "OnFamppExistsEvent() " 
         << " event:" << ev->getEventName() 
         << " filename:" << filename
         << endl;
}


void
OnFamppEndExistEvent( string filename, Fampp::fh_fampp_req req, Fampp::fh_fampp_ev ev )
{
    cerr << "OnFamppEndExistEvent() " 
         << " event:" << ev->getEventName() 
         << " filename:" << filename
         << endl;
}

void
OnFamppChangedEvent( string fqfilename, Fampp::fh_fampp_req req, Fampp::fh_fampp_ev ev )
{
    cerr << "OnFamppChangedEvent()"
         << " event:" << ev->getEventName() 
         << " filename:" << fqfilename
         << endl;
}

void
OnFamppDeletedEvent( string fqfilename, Fampp::fh_fampp_req req, Fampp::fh_fampp_ev ev )
{
    cerr << "OnFamppDeletedEvent()"
         << " event:" << ev->getEventName() 
         << " filename:" << fqfilename
         << endl;
}

void
OnFamppMovedEvent( string fqfilename, Fampp::fh_fampp_req req, Fampp::fh_fampp_ev ev )
{
    cerr << "OnFamppMovedEvent()"
         << " event:" << ev->getEventName() 
         << " filename:" << fqfilename
         << endl;
}


int 
main(int argc, char* argv[])
{
    try
    {

        // Hookup to glib for automatic handling during main loop
      Fampp::RegisterFamppWithGLib();

        string path = "/";
        if( argc > 1 )
            path = argv[1];
        // Create a request for /tmp
        cerr << "monitoring directory:" << path << endl;
        Fampp::fh_fampp_req req = Fampp::MonitorDirectory( path );

        // Connect libsigC++ signal
        req->getSig<Fampp::FamppCreatedEvent>(). connect(sigc::ptr_fun(&CreatedFamEv));
        req->getSig<Fampp::FamppExistsEvent>().  connect(sigc::ptr_fun(&OnFamppExistsEvent));
        req->getSig<Fampp::FamppEndExistEvent>().connect(sigc::ptr_fun(&OnFamppEndExistEvent));
        req->getSig<Fampp::FamppChangedEvent>(). connect(sigc::ptr_fun(&OnFamppChangedEvent));
        req->getSig<Fampp::FamppDeletedEvent>(). connect(sigc::ptr_fun(&OnFamppDeletedEvent));
        req->getSig<Fampp::FamppMovedEvent>(). connect(sigc::ptr_fun(&OnFamppMovedEvent));

        
        cerr << "=========================================" << endl;
        cerr << "=========================================" << endl;
        cerr << "=========================================" << endl;

//         while(1)
//         {
//             Fampp::Fampp::Instance().NextEvent();
//             cerr << "next ev" << endl;
//         }

//         while(1)
//         {
//             if( Fampp::Fampp::Instance().Pending() )
//                 Fampp::Fampp::Instance().NextEvent();
//         }
        
        GMainLoop *loop = g_main_loop_new(0,0);
        g_main_loop_run( loop );
        g_main_loop_unref( loop );

//		gtk_main();
		return 0;
    }
	catch( Fampp::FamppException& e )
    {
		cout << e << endl;
	}
}




src/Fampp2GlibSupport.cpp
/******************************************************************************
*******************************************************************************
*******************************************************************************

    Copyright (C) 2001 Ben Martin

    This file is part of fampp2.

    fampp2 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    fampp2 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with fampp2.  If not, see <http://www.gnu.org/licenses/>.

    For more details see the COPYING file in the root directory of this
    distribution.

    $Id: Fampp2GlibSupport.cpp,v 1.2 2010/09/24 05:16:50 ben Exp $

*******************************************************************************
*******************************************************************************
******************************************************************************/

#include <Fampp2GlibSupport.hh>
#include <glib.h>

#include <iostream>
using namespace std;


namespace Fampp
{
    ExceptionList_t ExceptionList;
    
    ExceptionList_t getExceptions()
    {
        return ExceptionList;
    }
    void
    clearExceptions()
    {
        ExceptionList.clear();
    }
    
    bool
    haveMoreExceptions()
    {
        return !ExceptionList.empty();
    }
    
    std::exception
    getNextException()
    {
        std::exception r = ExceptionList.front();
        ExceptionList.pop_front();
        return r;
    }
    
    
    void
    addException( const exception& e )
    {
        ExceptionList.push_back(e);
    }
    
    static gboolean
    registerWithGLib_cb( GIOChannel *source,
                         GIOCondition condition,
                         gpointer data)
    {
        while( Fampp::Instance().Pending() )
        {
            try
            {
                Fampp::Instance().NextEvent();
            }
            catch( std::exception& e )
            {
                cerr << "fampp2 exception: " << e.what() << endl;
                // Cant throw through glib callback
                // throw;
                addException( e );
            }
        }
        return 1; // call again
    }
    
    void
    ProcessFamppEvents()
    {
        while( Fampp::Instance().Pending() )
        {
            Fampp::Instance().NextEvent();
        }
    }
    
    

    void
    RegisterFamppWithGLib()
    {
        static bool alreadyRegistered = false;
        
        if( !alreadyRegistered )
        {
            alreadyRegistered = true;
        
            guint result;
            GIOChannel *channel;
            GIOCondition cond = GIOCondition(G_IO_IN | G_IO_ERR | G_IO_PRI);

//            cerr << "Attaching to fd:" << Fampp::Instance().getFD() << endl;
            
            channel = g_io_channel_unix_new( Fampp::Instance().getFD() );
            result  = g_io_add_watch( channel, cond,
                                      registerWithGLib_cb, 0 );
            g_io_channel_unref (channel);
        
        }
    }
 
};

